src/docs/runtime-metadata.md

```diff
@@ -0,0 +1,255 @@
+{
+    const nint kMaxDiscriminator = 0x7ff;
+    NativeHandle handle;
+
+    internal TypeMetadata (NativeHandle handle) // internal to help prevent bogus handles
+    {
+        this.handle = handle;
+    }
+
+    public NativeHandle Handle {
+        get { return handle; }
+    }
+
+    public TypeMetatypeKind {
+        get { return TypeMetadataKindFromHandle (handle); }
+    }
+
+    public static bool TryGetTypeMetadata (Type type, [NotNullWhen (true)] out TypeMetadata? result)
+    {
+        // implementation tbd
+    }
+
+    public static bool TryGetTypeMetadata<T> (T t, [NotNullWhen (true)] out TypeMetadata? result)
+    {
+        return TryGetTypeMetadata (typeof (t), out result);
+    }
+
+    static TypeMetadataKind TypeMetadataKindFromHandle (NativeHandle h)
+    {
+        nint val = ReadNativeInt (h.Handle);
+        if (val == 0)
+            return TypeMetadataKind.None;
+        if (val > kMaxDiscriminator)
+            return TypeMetadataKind.Class;
+        return (TypeMetadataKind)val;
+    }
+}
+```
+
+While not a complete implementation, this makes the core clear: this should be an immutable struct that can easily by passed by value to pinvokes and through reverse pinvokes.
+
+In addition, there should be the following **public** methods at a minimum:
+
+```csharp
+public bool TryGetNominalTypeDescriptor ([NotNullWhen (true)] out NominalTypeDescriptor? result) { }
+public string TypeName { get; } // uses the nominal type descriptor
+// NB - while Swift's Optional<T> is declarad as a generic enum, its implementation is a special case
+// since it is very common.
+public bool IsGeneric { get; } // uses nominal type descriptor
+public int GenericArgumentCount { get; }
+```
+
+In addition, we are likely to need **internal** accessors for:
+- the class superclass `TypeMetadata`
+- the number of elements in a tuple
+- the `TypeMetadata` of each tuple element
+- the number of parameters in a closure
+- the `TypeMetadata` of each parameter
+- the calling conventions of a closure
+- whether or not the closure has a return value
+- the `TypeMetadata` of the return value
+- whether or not parameters have flags (this determines if a parameter is `inout`, variadic or autoclosure)
+
+In addition, while we are not likely to need but is present:
+- Protocol information (flags, number of witness tables, protocol descriptors for each, whether a protocol is ObjC)
+
+## Mapping C# Type to TypeMetadata and Vice Versa
+
+There should be a threadsafe 1 to 1 bi-map of Type <-> TypeMetadata to serve as cache.
+
+For the Type -> TypeMetadata half, there should be a strategy mechanism that handles the following cases:
+- primitive value types such as int, uint, short, ushort etc. NB: IntPtr is a synonym for nint and UIntPtr is a synonym for nuint
+- bound Swift types should each have a static method for getting the type metadata using the metadata accessor which should be inspectable via reflection or interface compliance
```

Comment:
Is the reflection supported on static abstract interfaces?

Reply:
Yes. I did a prototype of it - the method name is unpronounceable in C#, but completely pronounceable in reflection.
```csharp
public interface IMeta {
    static abstract TypeMetadata GetTypeMetadata ();
}

// ...
    public static nint MetadataOf<T>(T t)
    {
        var tt = typeof(T);
        if (typeof(IF).IsAssignableFrom (tt)) {
            var mi = tt.GetMethod ("IFace. IMeta.GetTypeMetadata", BindingFlags.NonPublic | BindingFlags.Static);
            return (nint)mi.Invoke (null, null);
        }
        throw new NotImplementedException ();
    }
```